### PR TITLE
#3061 - Fix NPE when Measure refers to non-existent library

### PR DESCRIPTION
**Pull request by csandersdev**
I applied a fix before the holiday that corrected a category of NPE
errors related to execution against measure or library resources that
don't exist. It turns out that wasn't the core of the issue 3061. This
goes back and addresses when a valid measure reference is used, but that
Measure refers to a Library resource that does not exist. I also added
some handling that allows resolution of Library resources by reference
in addition to by canonical URL in the registry and also updated the
error handling around retrieving the primary library reference from the
Measure resource.

Signed-off-by: Corey Sanders <corey.thecolonel@gmail.com>

### CODE REVIEW

**Review by lmsurpre — APPROVED** [review]

File: cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/CareGapsOperation.java [hunk 1]
```diff
@@ -98,16 +98,16 @@ protected Parameters doInvoke(FHIROperationContext operationContext, Class<? ext
         }
     }
 
-    protected Bundle processAllMeasures(FHIRBundleCursor cursor, String subject, ZoneOffset zoneOffset, Interval measurementPeriod, FHIRResourceHelpers resourceHelper, TerminologyProvider termProvider, Map<String,DataProvider> dataProviders) {
+    protected Bundle processAllMeasures(FHIRBundleCursor cursor, String subject, ZoneOffset zoneOffset, Interval measurementPeriod, FHIRResourceHelpers resourceHelper, TerminologyProvider termProvider, Map<String,DataProvider> dataProviders) throws FHIROperationException {
         Bundle.Builder reports = Bundle.builder().type(BundleType.COLLECTION); 
         
         AtomicInteger count = new AtomicInteger(0);
-        cursor.forEach(resource -> {
+        for(Object resource : cursor) {
```
Comment (prb112) [inline cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/CareGapsOperation.java:105]:
```suggestion
        for (Object resource : cursor) {
```

File: cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/OperationHelper.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        List<org.cqframework.cql.elm.execution.Library> result =
+                libraries.stream().flatMap(fl -> LibraryHelper.loadLibrary(translator, fl).stream()).filter(Objects::nonNull).collect(Collectors.toList());
+        return result;
+    }
+    
+    public static Measure loadMeasureByReference(FHIRResourceHelpers resourceHelper, String reference) throws FHIROperationException {
+        return loadResourceByReference(resourceHelper, ResourceType.MEASURE, Measure.class, reference);
+    }
+    
+
```
Comment (prb112) [inline cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/OperationHelper.java:64]:
```suggestion

```

File: cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/OperationHelper.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    public static Measure loadMeasureById(FHIRResourceHelpers resourceHelper, String reference) throws FHIROperationException {
+        return loadResourceById(resourceHelper, ResourceType.MEASURE, reference);
+    }
+    
+    public static Library loadLibraryByReference(FHIRResourceHelpers resourceHelper, String reference) throws FHIROperationException {
```
Comment (prb112) [inline cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/OperationHelper.java:69]: javadocs for consistency would be great.

File: cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/OperationHelper.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        return loadResourceByReference(resourceHelper, ResourceType.LIBRARY, Library.class, reference);
+    }
+    
+
```
Comment (prb112) [inline cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/OperationHelper.java:73]:
```suggestion

```

File: cql/operation/fhir-operation-cqf/src/test/java/com/ibm/fhir/operation/cqf/BaseDataRequirementsOperationTest.java [hunk 5]
```diff
@@ -96,9 +100,11 @@ protected Library initializeLibraries(FHIRRegistry mockRegistry, FHIRResourceHel
         Library primaryLibrary = getPrimaryLibrary(cql, modelInfo, fhirHelpers, sde);
         List<Library> fhirLibraries = Arrays.asList(primaryLibrary, getSupplementalDataElementsLibrary(), getFHIRHelpers(), getFHIRModelInfo());
     
-        when(resourceHelper.doRead(eq("Library"), eq(primaryLibrary.getId()), anyBoolean(), anyBoolean(), any())).thenAnswer(x -> TestHelper.asResult(primaryLibrary));
-        
-        fhirLibraries.stream().forEach( l -> when(mockRegistry.getResource( canonical(l.getUrl(), l.getVersion()).getValue(), Library.class )).thenReturn(l) );
+        if( exists ) {
```
Comment (prb112) [inline cql/operation/fhir-operation-cqf/src/test/java/com/ibm/fhir/operation/cqf/BaseDataRequirementsOperationTest.java:103]:
```suggestion
        if (exists) {
```

File: cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/AbstractMeasureOperation.java [hunk 6]
```diff
@@ -86,19 +79,16 @@ public RetrieveProvider getRetrieveProvider(FHIRResourceHelpers resourceHelper,
         return retrieveProvider;
     }
 
-    public MeasureReport.Builder doMeasureEvaluation(Measure measure, ZoneOffset zoneOffset, Interval measurementPeriod, String subjectOrPractitionerId,
+    public MeasureReport.Builder doMeasureEvaluation(FHIRResourceHelpers resourceHelpers, Measure measure, ZoneOffset zoneOffset, Interval measurementPeriod, String subjectOrPractitionerId,
```
Comment (punktilious) [inline cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/AbstractMeasureOperation.java:82]: javadoc

**Review by prb112 — APPROVED** [review]
LGTM, copyright updates (as Lee mentioned)

File: cql/fhir-quality-measure/src/main/java/com/ibm/fhir/ecqm/r4/MeasureHelper.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+package com.ibm.fhir.ecqm.r4;
+
+import com.ibm.fhir.exception.FHIROperationException;
+import com.ibm.fhir.model.resource.Measure;
+
+/**
+ * Utility methods for working with Measure resources
+ */
+public class MeasureHelper {
+    public static String getPrimaryLibraryId(Measure measure) throws FHIROperationException {
```
Comment (punktilious) [inline cql/fhir-quality-measure/src/main/java/com/ibm/fhir/ecqm/r4/MeasureHelper.java:15]: javadoc

File: cql/fhir-quality-measure/src/main/java/com/ibm/fhir/ecqm/r4/MeasureHelper.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        String primaryLibraryId = null;
+        if( measure.getLibrary() != null && measure.getLibrary().size() == 1 ) {
```
Comment (punktilious) [inline cql/fhir-quality-measure/src/main/java/com/ibm/fhir/ecqm/r4/MeasureHelper.java:17]:
```
if (measure.getLibrary() ...
```

File: cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/CareGapsOperation.java [hunk 9]
```diff
@@ -98,16 +98,16 @@ protected Parameters doInvoke(FHIROperationContext operationContext, Class<? ext
         }
     }
 
-    protected Bundle processAllMeasures(FHIRBundleCursor cursor, String subject, ZoneOffset zoneOffset, Interval measurementPeriod, FHIRResourceHelpers resourceHelper, TerminologyProvider termProvider, Map<String,DataProvider> dataProviders) {
+    protected Bundle processAllMeasures(FHIRBundleCursor cursor, String subject, ZoneOffset zoneOffset, Interval measurementPeriod, FHIRResourceHelpers resourceHelper, TerminologyProvider termProvider, Map<String,DataProvider> dataProviders) throws FHIROperationException {
```
Comment (punktilious) [inline cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/CareGapsOperation.java:101]: javadoc nice to have (must have for public methods)

File: cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/OperationHelper.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+import com.ibm.fhir.exception.FHIROperationException;
+import com.ibm.fhir.model.resource.Library;
+import com.ibm.fhir.model.resource.Measure;
+import com.ibm.fhir.model.resource.Resource;
+import com.ibm.fhir.model.type.code.ResourceType;
+import com.ibm.fhir.persistence.SingleResourceResult;
+import com.ibm.fhir.registry.FHIRRegistry;
+import com.ibm.fhir.server.spi.operation.FHIRResourceHelpers;
+
+public class OperationHelper {
```
Comment (punktilious) [inline cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/OperationHelper.java:28]: javadoc

File: cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/OperationHelper.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    public static List<org.cqframework.cql.elm.execution.Library> loadCqlLibraries(List<Library> libraries) {
+        FHIRLibraryLibrarySourceProvider sourceProvider = new FHIRLibraryLibrarySourceProvider(libraries);
+        CqlTranslationProvider translator = new InJVMCqlTranslationProvider(sourceProvider);
+
+        List<org.cqframework.cql.elm.execution.Library> result =
+                libraries.stream().flatMap(fl -> LibraryHelper.loadLibrary(translator, fl).stream()).filter(Objects::nonNull).collect(Collectors.toList());
+        return result;
+    }
+    
+    public static Measure loadMeasureByReference(FHIRResourceHelpers resourceHelper, String reference) throws FHIROperationException {
```
Comment (punktilious) [inline cql/operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/OperationHelper.java:60]: javadoc

File: cql/operation/fhir-operation-cqf/src/test/java/com/ibm/fhir/operation/cqf/BaseDataRequirementsOperationTest.java (hunk 5 above)
Comment (punktilious) [inline cql/operation/fhir-operation-cqf/src/test/java/com/ibm/fhir/operation/cqf/BaseDataRequirementsOperationTest.java:103]:
```
if (exists)
```

**Pull request merged**